presto-main/src/main/java/com/facebook/presto/cost/CostComparator.java

```diff
@@ -59,6 +60,11 @@ public int compare(Session session, PlanNodeCostEstimate left, PlanNodeCostEstim
         requireNonNull(right, "right is null");
         checkArgument(!left.hasUnknownComponents() && !right.hasUnknownComponents(), "cannot compare unknown costs");
 
+        long queryMaxMemory = getQueryMaxMemory(session).toBytes();
+        if (left.getIndividualMemoryCost() > queryMaxMemory || right.getIndividualMemoryCost() > queryMaxMemory) {
```

Comment:
to me it is like hack that you are changing generic `compare` method. I would prefer to have a dedicated method to handle the case high memory consumption plan nodes.